common/client/send_only_node.go

```diff
@@ -0,0 +1,180 @@
+package client
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"sync"
+
+	"github.com/smartcontractkit/chainlink/v2/common/types"
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+	"github.com/smartcontractkit/chainlink/v2/core/utils"
+)
+
+type SendOnlyClient[
+	CHAIN_ID types.ID,
+] interface {
+	Close()
+	ChainID(context.Context) (CHAIN_ID, error)
+	DialHTTP() error
+}
+
+// SendOnlyNode represents one node used as a sendonly
+type SendOnlyNode[
+	CHAIN_ID types.ID,
+	RPC SendOnlyClient[CHAIN_ID],
+] interface {
+	// Start may attempt to connect to the node, but should only return error for misconfiguration - never for temporary errors.
+	Start(context.Context) error
+	Close() error
+
+	ConfiguredChainID() CHAIN_ID
+	RPC() RPC
+
+	String() string
+	// State returns nodeState
+	State() nodeState
```

Comment:
It's not possible to inject custom implementation of `SendOnlyNode|Node` as `nodeState` is private. 
Even if it's not needed, it feels weird

Reply:
Not sure I understand this. Even if `nodeState` is changed internally, `State()` is used by the MultiNode to access the state. Let me know if you have something else on your mind.

Reply:
During integration of new chain someone might benefit from possibility to create their own implementation of `SendOnlyNode` or `Node` and still use it with MultiNode. But it's not possible to create custom implementation in different package as there is no way to implement `State() nodeState` method as `nodeState` is private for `common/client` package. 

One way to resolve it is to make `SendOnlyNode` and `Node` private types and create instances of them in `NewMultiNode` using passed RPCs.

Another is to keep nodeState public

Reply:
Personally, I would go in the exact opposite direction and totally abstract Node from the user. Then they would only have to instantiate the MultiNode and the RPCs. The aim of this package is for the user to use the exact same API as Geth or other clients, without worrying about the complexity of internal node states. The only issue I see with that is that currently Node instantiation is coupled with `toml.Node` config and it might need some additional changes.

Reply:
I'm having trouble following this thread, but there are two different concerns here:
1. Whether or not the method is exported affects who can call it.
2. Whether or the returned type is exported affects who can _implement_ it.

Isn't it critical for this to be implemented by types outside of this package?

Reply:
> The aim of this package is for the user to use the exact same API as Geth or other clients, without worrying about the complexity of internal node states.

Can you elaborate on this? That doesn't fit my mental model. The Geth API is the thing that we want to abstract away, in order to reuse all the stuff that is not specific to geth's API, with other chain clients. Aren't the 'user's of this package the developers integrating those other chain clients? Do they need to implement this method? Call it?

Reply:
Maybe there are two separate uses here? Unexported & strongly-typed, and exported string?
```suggestion
	state() nodeState
```
```go
func (s *sendOnlyNode[CHAIN_ID, RPC]) State() string {
	return string(s.state())
}
```

Reply:
Yes, we do abstract the geth-specific calls into more generic ones, but at the same time, we should aim to abstract the Node as much as possible as well. I see two use cases here:
1. If a user wants to interact with a chain, instead of utilizing a Geth library or some other chain-specific one, we give them a slightly more generalized version of it that can apply to most chains. We should still aim to provide familiar methods like `BlockByHash` because the average user just wants to have methods as close as possible to the chain client.
2. To me the user shouldn't be worried about the internal node lifecycle and ideally about the creation and implementation of the node. The transition should be as simple as: "Instead of using a generalized API that makes direct calls to an RPC, you can pass multiple URLs and still use the same calls and we will take care of how the multi-node spins up nodes and handles them. Ideally, the nodes should be totally abstracted.
One example I can think of is the RTSP team. They want to have a client that interacts with a chain without worrying about a node's lifecycle and state.

Reply:
Ok, I think the problem starts from the fact that the original code exports all these fields and methods we are trying to generalize because they are used in different packages for mocking and testing (perhaps it's also on purpose to allow users to extend?).
So when we transferred everything we kept the same convention, but `common/client` doesn't have tests on the same package as it requires concrete types, and that means we could unexport everything and let the MultiNode act as the interface with the user.
But the question here is do we want the nodes to be extended by the user or totally abstracted? Also, something to consider is that once we agree on a strategy to test generalized packages we might end up going back to exporting those methods and fields.